### ISSUE TITLE
Using a GitHub Action instead of the Azure Function

### ISSUE THREAD

**Issue by softworkz** [gitgitgadget/gitgitgadget#809]
I was looking for a way to avoid setting up the Azure Function in addition to all the other components that are already part of this setup, so I was hoping to find a way that is using some of the existing components.

The solution that I'm showing is a small GitHub Action workflow and the obvious drawback is that it needs to be included in the repository files, that's unavoidable and the GitHub roadmap indicates that this won't change shortly.

Here's the workflow file (main.yml)

```yml
# This workflow invokes an Azure Pipeline for certain issue_comment events

name: gitgitgadget

# Controls when the workflow will run
on:
  # Triggers the workflow on push or pull request events but only for the master branch
  issue_comment:
    types: [created, edited]
# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "build"
  build:
    if: ${{ github.event.comment.user.type != 'Bot' }} && ${{ github.event.issue.pull_request }}
    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      - name: GitHub Script
        uses: actions/github-script@v5.0.0
        id: stringify
        with:
          script: |
            console.log(context.payload)
            return Buffer.from(JSON.stringify(context.payload)).toString('base64')
          result-encoding: string
      - name: Trigger Azure Release Pipeline
        uses: alexrster/pipelines@v2
        with:
          # Fullyqualified URL to the Azure DevOps organization along with project name
          # The secret needs to be the 'organization/project' part on DevOps
          azure-devops-project-url: https://dev.azure.com/${{ secrets.azure_devops_org_project }}

          # Name of the Azure Pipline to be triggered
          azure-pipeline-name: ${{ secrets.azure_devops_pipeline }}

          # Set/Overwrite pipeline variables
          azure-pipeline-variables: '{ "pr.comment.id": "${{ github.event.comment.id }}", "sourceBranch": "refs/pull/${{ github.event.issue.number }}/head", "payload": "${{ steps.stringify.outputs.result }}" }'

          # Paste personal access token of the user as value of secret variable:AZURE_DEVOPS_TOKEN
          azure-devops-token: ${{ secrets.azure_devops_token }}
```

**Comment by softworkz** [comment]
Let's compare a bit:

- The Azure function, can be called by anybody without authenticating and needs to perform checks to verify all requests
  The GH Action is not "callable" by anybody and doesn't need such checks
- The Azure function gets triggered by the gitigitgadget app which everybody can install. This requires checks to make sure that requests are only processes for the legitimate repos
  The GH Action is not called by the GH App. It directly works on a repository trigger. No checks required
- The Azure Function has checks to ignore certain events
  The GH Action doesn't needs this, it is only activated on the 'issue_comment' event, so it doesn't need that check either
- Both are checking for whether the comment event is about a PR
- The Azure function doesn't check for its own comments
  The GH Action checks `comment.user.type != 'Bot'` to avoid that the app doesn't get triggered from its own replies
- The Azure function doesn't check the event_type
  The GH Action gets only executed when [created, edited]

In summary, the GH Action invokes the Azure Pipeline even less often than the Azure function. Practically more or less equivalent in this regard.

I haven't come to a conclusion yet. I'm still exploring..

**Comment by softworkz** [comment]
## The GH Action does not appear to be a suitable alternative

I have abandoned the GitHub Action solution!

There are three reasons:

- Activation is too slow and that hits hard on the response times
- I realized, that as long as the action is not committed to the repo, the only way for having it, would be to rebase on each repo sync and let it "swim at the top of the head"
- This solution does not provide enough freedom and flexibility in the long run
  (even though, the checks could probably have been made visible by using the app for authentication, but it could never appear to be agile and accurate in timing - I mean at what points it is shown and completed).

## Response Time

I made some tests regarding response time as I think that's one of the most important acceptance criteria.
The test was always the same:
- Add a comment containing `/test`
- Measure the time between clicking the submit button and the arrival of the response
- That's the "Total Duration" column

Then I looked up the durations in the pipelines' history:

- Pipeline duration is the total duration for the pipeline
- Job duration is the duration of the job

Finally, I calculated the delay (overhead) of the first instance (AZ function, etc) as total duration minus pipeline duration
There are negative durations because I measured the time until the response appears in the browser, while the jobs are usually running a little longer

## Results

| Impl           | Run | Total Duration | DevOps Pipeline Duration | DevOps Job Duration | Function Overhead |
|----------------|----:|---------------:|-------------------------:|--------------------:|------------------:|
| Azure Function |   1 |            46s |                      34s |                 22s |               12s |
|                |   2 |            30s |                      27s |                 15s |                3s |
|                |   3 |            28s |                      26s |                 16s |                2s |
|                |     |                |                          |                     |                   |
| GitHub Action  |   1 |            45s |                      27s |                  6s |               18s |
|                |   2 |            40s |                      26s |                  5s |               14s |
|                |   3 |            47s |                      29s |                  6s |               18s |
|                |     |                |                          |                     |                   |
| Self Hosted    |   1 |            23s |                      28s |                  6s |               -5s |
|                |   2 |            22s |                      24s |                  9s |               -2s |
|                |   3 |            26s |                      27s |                  6s |               -1s |

## Analysis

At the top we have the current gitgitgadget behavior. Next are the values from my GitHub action and the last is what I have chosen eventualy.

### Azure Function

What can be seen with the Azure function is that there's quite some delay at the first invocation  which is decreasing on subsequent invocations. The subsequent values would be OKl, but the nature of this application is pattern that doesn't fit well for this hosting model: the function is hit only infre4quent, which means that it will regularly expose that delay (always after some period of inactivity. 

### GiitHub Action

The table shows it: the amount of delay that the Action adds, is way too high.

What's also visible, though is the runtime reduction of the job that I was able to achieve like explained elsewhere.
Those 6s are pretty consistent. Much lower than the ggg times. What's odd is that in my cases, the time for the pipeline runs without job is much higher than in your case. 
I hope this is causes by the current degradation of their Azure infrastructure  In your case the main duration is 10-12s, in my case it's 18-20s. If my suspicion would stand, I could probably achieve response times of about 15-18s.

### Self-Hosted Endpoint

I have set up an API endpoint on the build agent server via nodejs/nginx. It needed some smaller changes to the Azure function code, but it's essentially the same.

**Comment by dscho** [comment]
> The Azure function, can be called by anybody without authenticating and needs to perform checks to verify all requests

Well, authentication has to happen _somewhere_. And I don't think that [the `validateGitHubWebHook()` function](https://github.com/gitgitgadget/gitgitgadget/blob/2551a46dfd77a2a4cba7768705524f9406afb52d/azure-function/index.js#L15-L35) is complicated at all.

> The GH Action doesn't needs this, it is only activated on the 'issue_comment' event, so it doesn't need that check either

Yes, we need to verify that we're actually looking at a GitGitGadget command, otherwise we'd burn up minutes unnecessarily just to process a regular (and legitimate) PR comment.

FWIW I had expected a proposal not so much to just replace the Azure Function by a GitHub workflow, but instead to replace _both_ the Azure Function _and_ the Azure Pipeline by a GitHub workflow.

And it might still be possible to do that, but as you say, it would require a code change in the target repository, which is making things a lot harder (especially once you need the flexibility to debug problems by editing the Pipeline and then re-running failed jobs).

**Comment by softworkz** [comment]
Exactly - and then updates: you'll end up having different action workflows in different branches. So you'd probably still end up just having a small (rarely changing) stub in the actual repos and try to use have the workflow definitions elsewhere.

GH have the perfect solution for this on their roadmap: "Organization-Workflows"
But it's getting postponed regularly.

**Comment by dscho** [comment]
Hmm. You've got me thinking, and I start to believe that https://docs.github.com/en/actions/learn-github-actions/reusing-workflows is what we are looking for.

Of course, we would then still have to guard the job by a condition that checks the following:
- the necessary secret variables are configured
- the payload (read: the comment) starts with a slash. We might even be able to use [`!contains(payload...comment, "\n")`](https://docs.github.com/en/actions/learn-github-actions/expressions#contains) to verify that it is a one-liner.

**Comment by softworkz** [comment]
I have new numbers today as MS have moved my DevOps hosting the US (on my request)

| Impl           | Run | Total Duration | DevOps Pipeline Duration | DevOps Job Duration | Function Overhead |
|----------------|----:|---------------:|-------------------------:|--------------------:|------------------:|
| Self Hosted    |   1 |            17s |                      18s |                  5s |               -1s |
| and Azure        |   2 |            17s |                     19s |                  7s |               -2s |
|  in the US          |   3 |            18s |                     20s |                  5s |               -2s |

Now it's becoming really  nice to interact with..

**Comment by softworkz** [comment]
> Hmm. You've got me thinking, and I start to believe that [docs.github.com/en/actions/learn-github-actions/reusing-workflows](https://docs.github.com/en/actions/learn-github-actions/reusing-workflows) is what we are looking for.
> 
> Of course, we would then still have to guard the job by a condition that checks the following:
> 
> * the necessary secret variables are configured
> * the payload (read: the comment) starts with a slash. We might even be able to use [`!contains(payload...comment, "\n")`](https://docs.github.com/en/actions/learn-github-actions/expressions#contains) to verify that it is a one-liner.

My latest idea is heading into a slightly different. direction. Concurrency needs to be ensured, but I think this can always be handled in one ore another way and no matter where (and how) a job is run.
But that's not the case for the app interactions. Even though I have nice response time now in testing, that won't we for granted constantly over time. Especially not, when we have a single-runner execution strategy, because that means that while one of the rather heavy-lifting jobs is running, it would block all the others which need to be responsive.

That's why I'm thinking about a split of the jobs into two separate groups.

That split would separate the operations, which are lightweight and short-lived from those other jobs that take longer longer to execute. Another line of decision would be about which jobs need a fully checked out and persistent repo and which don't or for which there aren't any time constraints regarding execution duration. 

I'm far from suggesting though. that everything should be reworked and done differently. I had chosen to get involved and adopt this because I like how most things are done. I don't have much time to rework things anyway.  So, my mission is more about taking a deep, and intensive but step into into this an and getting the maximum out of it, that can be achieved without deviating from the established and proven patterns and taking things at risk

The new idea is going along those lines. What we have already seen is that neither  GH actions nor Azure jobs are a an ideal choice for driving an interactive GitHub app . There's quite a range of different reasons involved and they are all summing up to those delays that we are seeing.
Just naturally, I've thought for a moment about getting rid of Azure and actions and taking care about the execution automation ourselves. But that would be painful, take quite a while and effort until it's stable enough and then it would still be far from perfect. 
But the actual idea might come close: 

1. We install Azure DevOps Server Express on the buil()d agent server
   It's identical in features to the regular one with th exception that it's limited to 5 users. 
   2 will probaby suffice
2. We register the build runner to the local DevOps server and remove it from the other cloud agent pool
3. We move the lightweight pipelines to our own DevOps server and lever the others on cloud-Azure

Almost no changes will be required for this,
As a result, we'd have a DevOps server which is serving a single client only --- us!
And it has a single single, fixed-assigned agent. It doesn't need to any agent acquisitioning qany provisioning, it da,searching. browsing], etc. I'd 3xcü, '...

I'd expect substantial improvements from thsis - what do you think?

**Comment by dscho** [comment]
Well, whatever improvements we plan on making, I guess we should work on keeping separate concerns separate, so that e.g. `git/git` can stay with the current strategy, FFMPEG can run with what you suggest, and the core functionality is implemented identically.

**Comment by softworkz** [comment]
Totally agree. That's how I even came to the above: 
having a local DevOps server instead of the cloud service, would be a "neutral" deviation.

Also, I have abandoned the GitHub action and I'm running the same Azure function like you do (just not on Azure).
So right now, we're quite equal. 

Except for my script changes. I guess I'll need to factor them out as configurable options.

**Comment by softworkz** [comment]
> Hmm. You've got me thinking, and I start to believe that [docs.github.com/en/actions/learn-github-actions/reusing-workflows](https://docs.github.com/en/actions/learn-github-actions/reusing-workflows) is what we are looking for.
> 
> Of course, we would then still have to guard the job by a condition that checks the following:
> 
> * the necessary secret variables are configured

Thinking about a different kind of interaction, I just realized that this doesn't fly.

Every PR can contain changes to the workflows/actions as these are part of the repo content, and to make this work at all, we'd need to allow execution of workflows for PRs where the source is coming from a forked repo's branch. Such a PR could contain a modified workflow that prints out all secrets (or do whatever else with them).

(my action above would have been just moderately unsafe as it only triggered the DevOps pipelines, but could still be overwritten to disrupt functionality)

I think, before GH will introduce organization-actions, there's no way to go with Actions.
(hence closing)

**Comment by dscho** [comment]
> Every PR can contain changes to the workflows/actions as these are part of the repo content, and to make this work at all, we'd need to allow execution of workflows for PRs where the source is coming from a forked repo's branch. Such a PR could contain a modified workflow that prints out all secrets (or do whatever else with them).

That is almost all true.

However, the GitHub workflow would be started via a PR comment. Such a workflow's definition comes from the primary branch, always, and hence _has_ access to the secrets. Any workflow started from the branch does _not_ have access to the secrets.

One remaining question is how to make the workflow run show up in the PR's checks. I guess this would need to be done "manually".

**Comment by softworkz** [comment]
> Such a workflow's definition comes from the primary branch, always

Another problem is this: the repository regularly gets synchronized from the actual source repository (like git/git).
So, when you add a workflow to the (gitgitgadget/git) repository via a commit, that workflow would always disappear when it gets synced, which means that the workflow change needs to be made in the source repo (git/git).
And what comes on top of this: gitgitgadget/git is a already a FORK of git/git and as such, the restrictions of GitHub actions for forked repositories already apply when running the actions in gitgitgadget/git...
